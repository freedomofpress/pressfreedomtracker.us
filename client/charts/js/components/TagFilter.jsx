import React, { useContext, useState } from 'react'
import * as d3 from 'd3'
import { flatMap, countBy } from 'lodash'
import {
	TOGGLE_PARAMETER_ITEM,
} from '../lib/actionTypes'
import { FiltersDispatch } from '../lib/context'
import CheckBoxBar from './CheckBoxBar'
import AutoComplete from './Autocomplete'
import { trackMatomoEvent } from '../lib/utilities'

const margins = {
	top: 10,
	left: 50,
	right: 30,
	bottom: 50,
}

export default function TagFilter({
	dataset,
	width,
	initialPickedTags,
	filterParameters: selectedTags,
}) {
	const updateFilters = useContext(FiltersDispatch)

	// "Picked" tags are ones the user has chosen from the extended
	// tag list, or via the URL parameters.  These will always be
	// displayed with a checkbox in the list, regardless of incident
	// count.
	const [pickedTags, setPickedTags] = useState(Array.from(initialPickedTags))

	const tags = countBy(
		flatMap(dataset, 'tags')
	)

	const countTags = Object.entries(tags)
			.map(([tag, count]) => ({
				tag: tag,
				count: count,
			}))
			.sort((a, b) => b.count - a.count)

	const xScale = d3
			.scaleLinear()
			.domain([0, d3.max(countTags.map((d) => d.count))])
			.range([margins.left, width - margins.right - margins.left])

	// separate the tags into the top three (by incident count) and
	// the rest.
	let extendedTags = countTags.splice(3)

	// separate custom tags the user has picked from the rest into
	// their own list, the remove them from the rest.

	// Remove any picked tags from the extended tags. Picked tags are
	// already part of the check box list and should not appear in the
	// drop-down chooser.
	extendedTags = extendedTags.filter(({tag}) => !pickedTags.includes(tag))

	// Compute what tags to display in the check box list: the top 3
	// tags (by incident count) plus all picked tags.
	let topTags = countTags.map(({tag}) => tag)
	let displayTags = countTags.concat(
		// Picked tags minus any that are already in the top 3, then
		// listed alongside their incident count (with zero as a
		// default).
		pickedTags.filter(tag => !topTags.includes(tag)).map(tag => ({count: tags[tag] ?? 0, tag}))
	)
	return (
		<div className="filters__form--fieldset">
			{displayTags.map(({ tag, count }, i) => {
				const isSelected = selectedTags.has && selectedTags.has(tag)
				return (
					<CheckBoxBar
						key={i}
						width={width-40}
						label={`#${tag}`}
						count={count}
						barWidth={count > 0 ? xScale(count) : 0}
						isSelected={isSelected}
						index={i}
						onClick={() => {
							updateFilters({
								type: TOGGLE_PARAMETER_ITEM,
								payload: {
									filterName: 'tags',
									item: tag,
								}
							})
							// Fire matomo event
							trackMatomoEvent(['Filter', 'Tags', 'Toggle', tag])
						}}
					/>
				)
			})}
			<AutoComplete
				suggestions={extendedTags}
				suggestionsLabelField="tag"
				suggestionsSidenoteField="count"
				name="tags"
				placeholder="Find tags..."
				itemNamePlural="tags"
				itemNameSingular="tag"
				handleSelect={(tag) => {
					setPickedTags(
						(previousPickedTags) => [...previousPickedTags, tag]
					)
					updateFilters({
						type: TOGGLE_PARAMETER_ITEM,
						payload: {
							filterName: 'tags',
							item: tag,
						}
					})
					// Fire matomo event
					trackMatomoEvent(['Filter', 'Tags', 'Select', tag])
				}}
			/>
		</div>
	)
}
