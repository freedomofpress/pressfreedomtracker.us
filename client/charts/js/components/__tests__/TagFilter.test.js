import React from 'react'
import ShallowRenderer from 'react-test-renderer/shallow'
import TagFilter from '../TagFilter'

test('renders TagFilter with mocked data', () => {
	const renderer = new ShallowRenderer();
	expect(renderer.render(
		<TagFilter
			dataset={[
				{
					"date": "2023-02-12T00:00:00.000Z",
					"city": "Thomasland",
					"state": "ID",
					"latitude": null,
					"longitude": null,
					"categories": [
						"Arrest / Criminal Charge"
					],
					"tags": [
						"lemur"
					]
				},
				{
					"date": "2022-10-23T00:00:00.000Z",
					"city": "New Brooke",
					"state": "NC",
					"latitude": null,
					"longitude": null,
					"categories": [
						"Arrest / Criminal Charge"
					],
					"tags": [
						"baboon"
					]
				},
				{
					"date": "2022-05-12T00:00:00.000Z",
					"city": "East Mary",
					"state": "CT",
					"latitude": null,
					"longitude": null,
					"categories": [
						"Equipment Search or Seizure"
					],
					"tags": [
						"civet"
					]
				}
			]}
			width={200}
			initialPickedTags={new Set()}
			filterParameters={new Set()}
		/>
	)).toMatchSnapshot();
});
