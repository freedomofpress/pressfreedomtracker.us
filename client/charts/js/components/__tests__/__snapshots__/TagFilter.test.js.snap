// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders TagFilter with mocked data 1`] = `
<div
  className="filters__form--fieldset"
>
  <CheckBoxBar
    barWidth={120}
    count={1}
    index={0}
    isSelected={false}
    label="#lemur"
    onClick={[Function]}
    width={160}
  />
  <CheckBoxBar
    barWidth={120}
    count={1}
    index={1}
    isSelected={false}
    label="#baboon"
    onClick={[Function]}
    width={160}
  />
  <CheckBoxBar
    barWidth={120}
    count={1}
    index={2}
    isSelected={false}
    label="#civet"
    onClick={[Function]}
    width={160}
  />
  <AutoComplete
    handleSelect={[Function]}
    itemNamePlural="tags"
    itemNameSingular="tag"
    name="tags"
    placeholder="Find tags..."
    suggestions={[]}
    suggestionsLabelField="tag"
    suggestionsSidenoteField="count"
  />
</div>
`;
